### ISSUE TITLE
detailed, agent-ready plan for E8-S6: Metal attention components

### ISSUE THREAD

**Issue by principle-lgtm** [stevedores-org/oxidizedMLX#78]
Below is a detailed, agent-ready plan for E8-S6: Metal attention components that’s realistic given what you’ve already scaffolded (GEMM, RMSNorm/LN, RoPE). I’ll give you:

scope (what “attention components” means for this story)

minimal kernels + composition path

Rust API + file layout

“functional attention block” golden test design (Python generator + Rust runner)

acceptance criteria that are actually binary

upgrade path (fused/flash attention later)

This story should not try to do full FlashAttention yet. The goal is correctness + a composition framework.

E8-S6: Metal attention components (expanded)
Goal

Implement the minimal attention building blocks on Metal and demonstrate that:

masking + softmax + matmul compose correctly and efficiently enough to be useful

a single-head or small multi-head attention block matches a reference implementation via golden tests

You will later fuse/optimize, but E8-S6 proves the “attention pipeline” works end-to-end on GPU.

Scope (Phase 1)
Must support

Scaled dot-product attention (SDPA) for one block:

scores = (Q @ K^T) * scale

apply causal mask (or provided mask)

P = softmax(scores, axis=-1)

Y = P @ V

Constraints (for E8-S6)

dtype: FP16 inputs, FP32 accumulation inside softmax + scaling

shape: start with single head to simplify:

Q: [Tq, Dh]

K: [Tk, Dh]

V: [Tk, Dh]

output Y: [Tq, Dh]

mask: causal (j > i => -inf) OR optional “padding mask” buffer later

Not required (explicitly out of scope)

fused FlashAttention

multi-query attention (MQA/GQA)

KV cache layout optimizations

dropout

backprop/autograd

Deliverables
crates/mlx-metal/src/
  attention.rs
  kernels/
    softmax_masked_fp16.metal
    scale_add_mask_fp16.metal   (optional)
tests/
  attention_block_goldens.rs
tools/py_ref_runner/
  gen_attention_goldens.py
tools/goldens/attn/
  case_small/
  case_llm_small/

You already have GEMM; reuse it:

MetalContext::gemm() for QK^T and P@V

Add softmax kernel for scores -> P with masking

Component design
Component A: GEMM (already from E8-S3)

We need two GEMMs:

scores = Q @ K^T

out = P @ V

Because your GEMM likely supports only A @ B without transpose, do one of:

add a GEMM “B transposed” mode (preferred, small extension)

explicitly transpose K into Kt (extra kernel + buffer)

For E8-S6, I’d do (1) GEMM variant:

gemm(a, b, trans_b=true) with b read as B[j*K + k] instead of B[k*N + j] (depending on layout)

If you don’t want to extend GEMM yet, add a transpose kernel for K (you’ll need it anyway eventually).

Component B: Softmax + mask kernel (new)

Kernel input: scores [Tq, Tk] (fp16 or fp32 buffer)
Kernel output: P [Tq, Tk] (fp16)
Mask: causal computed from indices, no mask buffer needed

Steps per row i:

compute max over j <= i (causal) else ignore

compute sum exp(scores - max) over allowed j

write exp(scores-max)/sum for allowed j, else 0

Accumulate in fp32.

Rust API
attention.rs minimal entry points
pub struct AttnParams {
  pub tq: usize,
  pub tk: usize,
  pub dh: usize,
  pub scale: f32,      // usually 1/sqrt(dh)
  pub causal: bool,
}

pub fn attention_fp16(
  ctx: &MetalContext,
  q: &MetalBuffer<u16>,   // [tq, dh]
  k: &MetalBuffer<u16>,   // [tk, dh]
  v: &MetalBuffer<u16>,   // [tk, dh]
  out: &MetalBuffer<u16>, // [tq, dh]
  p: AttnParams,
) -> Result<()>;

Internals:

allocate scores buffer [tq, tk] (fp16 or fp32; fp16 is ok if you scale in kernel; fp32 improves stability)

allocate prob buffer [tq, tk] fp16

scores = q @ k^T (via GEMM)

prob = softmax_mask(scores, scale, causal)

out = prob @ v (via GEMM)

Metal kernel scaffold
softmax_masked_fp16.metal (Phase 1, causal + scaling)

Input scores: fp16

Output probs: fp16

Params: tq, tk, scale, causal

Dispatch: 1 threadgroup per row, 256 threads.

each thread scans j in strides

reduce max + sum in threadgroup shared memory

This is similar to your RMSNorm reduction pattern.

Functional attention block golden test
Golden generator (Python)

Generate:

Q, K, V random (float32)

run reference attention in numpy:

scores = q @ k.T * scale

apply causal mask: set scores[i, j>i] = -1e9 (or -inf)

softmax stable

out = probs @ v

cast output to fp16 bits and save .npy

You can optionally store intermediate goldens (scores, probs) for debugging; acceptance requires only final output.

Test cases

small: tq=4, tk=4, dh=16

medium: tq=32, tk=32, dh=64

llm-ish small: tq=128, tk=128, dh=128 (still manageable)

add one asymmetric case: tq=8, tk=16, dh=32 (non-square)

Rust test runner

loads q/k/v bit arrays into Metal buffers

runs attention_fp16

compares output bits to expected (fp16 bits exact) OR compare with tolerance in float32

Recommendation: use allclose tolerance for attention outputs rather than exact bits, because trig/exp reductions can differ slightly between CPU numpy and GPU kernel.
Use: atol=3e-2, rtol=2e-2 initially.

Acceptance Criteria (binary)

✅ E8-S6 is done when:

attention_fp16() exists and runs end-to-end on GPU

Golden tests pass for at least:

case_small

case_medium

Correctness vs reference:

output matches within tolerances (documented)

No panics; invalid shapes return Result::Err

“Compose efficiently” baseline:

a bench or timed run prints ms for case_medium and case_llm_small

(not performance-gated yet, but recorded)

Implementation checklist (agent-style)
Task 1 — Extend GEMM for K^T or add transpose kernel

Choose one path; document it

Add unit tests for transpose or transposed GEMM correctness

Task 2 — Implement softmax_masked_fp16 kernel + Rust dispatcher

Stable max reduction

Stable exp/sum reduction

Causal mask logic

Task 3 — Implement attention_fp16 composition

allocate temporary buffers

call GEMM, softmax kernel, GEMM

Task 4 — Golden generator + Rust tests

generate goldens

run tests (tolerances)

Task 5 — Perf baseline harness

at least one benchmark case prints timings (not strict gating)

Upgrade path (future stories)

Once E8-S6 is green, the next wins are:

fuse QK^T + softmax + PV into FlashAttention-like kernels

add KV-cache friendly layout

add multi-head batched kernels

add per-token incremental decode kernels
